feat(xo-server): remember server's pool name and description

b-Nollet opened the pull request:
### Description

Remember server's pool name and description, and display the pool name as server default label. The server label can be deleted (since 81e7e0731220cf17d58245d508582aa6c5534531) to get back to the default pool name value.

We did multiple commits to keep the original changes around, but I would advise not to review by commit.

[XO-437](https://project.vates.tech/vates-global/projects/70ab2907-1ac3-4e7d-831f-a8752c36474d/issues/473e6228-1a1d-40f4-87d6-9669b406faf5)

### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

### Review process

> This 2-passes review process aims to:
>
> - develop skills of junior reviewers
> - limit the workload for senior reviewers
> - limit the number of unnecessary changes by the _author_

1. The _author_ creates a PR.
2. Review process:
   1. The _author_ assigns the _junior reviewer_.
   2. The _junior reviewer_ conducts their review:
      - Resolves their comments if they are addressed.
      - Adds comments if necessary or approves the PR.
   3. The _junior reviewer_ assigns the _senior reviewer_.
   4. The _senior reviewer_ conducts their review:
      - If there are no unresolved comments on the PR → merge.
      - Otherwise, we continue with **3.**
3. The _author_ responds to comments and/or makes corrections, and we go back to **2.**

Notes:

1. The _author_ can request a review at any time, even if the PR is still a _Draft_.
2. In theory, there should not be more than one reviewer at a time.
3. The _author_ should not make any changes:
   - When a reviewer is assigned.
   - Between the _junior_ and _senior_ reviews.

b-Nollet committed:
feat(xo-server): remember server's pool name and description

pdonias commented on packages/xo-server/src/api/pool.mjs:
```diff
@@ -43,6 +43,10 @@ export async function set({
+  if (nameLabel !== undefined || nameDescription !== undefined) {
+    const serverId = this.getXenServerIdByObject(pool.uuid, 'pool')
+    await this.updateXenServer(serverId, { poolNameDescription: nameDescription, poolNameLabel: nameLabel })
+  }
```
A pool's name may be changed by another client. Also, I think the most common use case for this feature would be to get the name of the pool as soon as the user connects to the server. Updating it when the pool's name_label changes is almost a bonus because it doesn't change that often.

So I think a better approach for this would probably be a sort of hook on the XAPI events to catch any changes of the pools' names and update their servers accordingly.

b-Nollet replied:
What I did does display the name of the pool when a server is added or connected to.

But you're right, I did not realize the pool's name could be changed by another client.

julien-f replied:
Yes, you need to watch objects and update the record if there was a change.

A good place for this coold be: https://github.com/vatesfr/xen-orchestra/blob/93a11cfb28791daf9a4f3050b6ec27efac90dde5/packages/xo-server/src/xo-mixins/xen-servers.mjs#L219

fbeauchamp commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -340,6 +354,12 @@ export default class XenServers {
+        await this.updateXenServer(id, {
+          poolNameLabel: xapi.pool.name_label,
+          poolNameDescription: xapi.pool.name_description,
+        })
```
is it also possible to not update it if the value do not change  ? 
Are we sure the xapi.pool.name_xx are correctly loaded here ?

b-Nollet committed:
reverting previous changes

b-Nollet committed:
suggested changes

julien-f commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -223,6 +262,11 @@ export default class XenServers {
 
+      // save pool name and description in server properties
+      if (xapiObject.$type === 'pool') {
+        updatePoolServer(xapiObject, xapiId)
```
What happens if this function reject?

b-Nollet replied:
The function is not awaited, so if it rejects it does not prevent the rest of the code from executing.

julien-f replied:
Promise rejections should always be handled, either:

- `await` them
- log the error
- explicitly ignore it `.catch(noop)` or better `ignoreErrors.call(promise)` (which does not ignore programming errors)

pdonias commented on CHANGELOG.unreleased.md:
Don't forget to change `xo-server patch` to `xo-server minor` in "Packages to release".

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +231,29 @@ export default class XenServers {
+          poolNameDescription: xapiObject.name_description,
+        })
+      }
+    }.bind(this)
```
Why not using an arrow function?

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +231,29 @@ export default class XenServers {
+      const xenServer = await this.getXenServer(serverId)
+
+      const serverPropertiesUpdate = {
+        ...(xapiObject.name_label !== undefined &&
```
Can a pool's `name_label` be `undefined`?

b-Nollet replied:
I'm not sure, I'll have to check

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +231,29 @@ export default class XenServers {
+          }),
+        ...(xenServer.poolNameLabel === undefined &&
+          xapiObject.name_label !== undefined && { label: xapiObject.name_label }),
+      }
```
I find it easier to read with explicit `if`s and assignments:
```js
const serverPropertiesUpdate = {}
if (xapiObject.name_label !== xenServer.poolNameLabel) {
  serverPropertiesUpdate.poolNameLabel = xapiObject.name_label
}
// ...
```
What do you think?

b-Nollet replied:
I don't really like it as it uses a lot of lines for a simple purpose, but if you had trouble reading what I did then I guess you're right

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -155,6 +167,10 @@ export default class XenServers {
+    if (poolNameDescription !== undefined) server.poolNameDescription = poolNameDescription || undefined
+    if (poolNameLabel !== undefined) server.poolNameLabel = poolNameLabel || undefined
+    // using pool name as default server name
+    if (server.label === undefined) server.label = server.poolNameLabel
```
Then if the pool's `name_label` changes twice, the server's `label` will keep the pool's old name?

b-Nollet replied:
Yes, but I don't think we have much other choice if we still want the server `label` to be directly configurable. (we could maybe use a boolean to tell if the pool label has ever been directly modified, but it would look really ugly)

pdonias replied:
True, but maybe what we could do, then, is not edit the server's `label` directly but fallback to `poolNameLabel` [somewhere here](https://github.com/vatesfr/xen-orchestra/blob/master/packages/xo-server/src/xo-mixins/xen-servers.mjs#L555-L572) if the label is `undefined`. What do you think? /cc @julien-f

julien-f replied:
Yep, that's what I thought initially: store both the user label and the pool's name and make the label defaults to the pool's name if `undefined` or empty when returning from the API.

b-Nollet replied:
Ok, I wasn't really familiar with this function but it makes more sense to do it like this. I'll change that

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +231,34 @@ export default class XenServers {
+      }
+      if (
+        xapiObject.name_description !== xenServer.poolNameDescription &&
+        !(xapiObject.name_description === '' && xenServer.poolNameDescription === undefined)
```
I don't understand this condition, could you explain why it's not simply `xapiObject.name_description !== xenServer.poolNameDescription`?

b-Nollet replied:
When a pool has no description, we have `xapiObject.name_description === ''` and `xenServer.poolNameDescription === undefined`.
I added that part to the condition to avoid calling `updateXenServer({poolNameDescription: ''})` when it won't have any effect, otherwise it would be called every time `updatePoolServer` is called for that server.

(we can't just check that `xapiObject.name_description !== ''` because a user may clear the description of a pool)

pdonias replied:
Why wouldn't `updateXenServer({ poolNameDescription: '' })` have any effect? Doesn't it update it when we pass `''`? I don't see where the difference is between `name_label` and `name_description`.

b-Nollet replied:
I noticed it doesn't update when `''` is passed. I guess it might also be true for the name, I just don't know if we can have a pool without name

pdonias replied:
@julien-f Is this the expected behavior? When we update a server a we pass `''` for a field, it deletes that field?

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +229,34 @@ export default class XenServers {
+      }
+      if (xenServer.poolNameLabel === undefined) {
+        serverPropertiesUpdate.label = xapiObject.name_label
+      }
```
Since you're already exposing the `label` [here](https://github.com/vatesfr/xen-orchestra/pull/8206/files#diff-a66f1df3735f66aef97a3695c6a20cb909e709ad46b41aefc814e1be28617fecR617), you can remove this. Or did I miss something?

b-Nollet replied:
You're right, I forgot to remove that part

pdonias commented on packages/xo-server/src/xo-mixins/xen-servers.mjs:
```diff
@@ -215,6 +229,34 @@ export default class XenServers {
+        return this.updateXenServer(serverIdsByPool[xapiId], {
+          poolNameLabel: xapiObject.name_label,
+          poolNameDescription: xapiObject.name_description,
+        })
```
Didn't you mean to pass `serverPropertiesUpdate`?

b-Nollet committed:
suggested changes

julien-f committed:
alternate proposal

- make updateXenServer only update database on change
- call updateXenServer directly in _onXenAdd

julien-f committed:
fix

julien-f committed:
fix

julien-f committed:
fix

julien-f committed:
fix

b-Nollet commented:
After the last fix it seems to work great.

fbeauchamp committed:
Merge branch 'master' into remember-pool-name
